benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java

```diff
@@ -221,281 +245,115 @@ private synchronized void handleTestRunningState(String testID) {
-
-    // TODO - should go into a stateless queue (so that we can recover)
-    Future<Boolean> future = taskExecutorService.submit(task);
-
-    // replace with new task
-    testTasks.put(testID, future);
-
-    try {
-
-      boolean acceptablePredictionError = future.get();
-
-      if (acceptablePredictionError) {
-
-        testModelDAO.setTestState(testID, TERMINATED);
-        testModelDAO.setTestTerminatedState(testID, GOAL_REACHED);
-
-        // no need to execute further
-        testTasks.remove(testID);
-
-      } else {
-
-        testModelDAO.setTestRunningState(testID, REMOVE_NON_REACHABLE_EXPERIMENTS);
-
-        handleTestState(testID);
-      }
-
-    } catch (InterruptedException | ExecutionException e) {
-      // TODO - handle  properly
-      e.printStackTrace();
-    } catch (BenchFlowTestIDDoesNotExistException e) {
-      // should not happen since it was added earlier
-      logger.error("test ID does not exist - should not happen");
-    }
-  }
-
-  private void removeNonReachableExperiments(String testID) {
-
-    logger.info("removeNonReachableExperiments with testID: " + testID);
-
-    RemoveNonReachableExperimentsTask task = new RemoveNonReachableExperimentsTask(testID);
+  private void handleTerminatedState(String testID) {
+    // remove max running time timeout
+    ScheduledFuture timeoutTaskFuture = timeoutTasks.remove(testID);
+    timeoutTaskFuture.cancel(true);
 
-    // TODO - should go into a stateless queue (so that we can recover)
-    Future<?> future = taskExecutorService.submit(task);
+    // remove any tasks left
+    testTasks.remove(testID);
 
-    // replace with new task
-    testTasks.put(testID, future);
-
-    waitForRunningTaskToComplete(testID, future, DETERMINE_EXECUTE_EXPERIMENTS);
+    // remove test from running queue so dispatcher can run next test
+    runningQueue.remove(testID);
   }
 
-  private void validateTerminationCriteria(String testID) {
-
-    logger.info("validateTerminationCriteria with testID: " + testID);
 
-    ValidateTerminationCriteria task = new ValidateTerminationCriteria(testID);
+  /**
+   * Called when user terminates test or max run time has been reached.
+   */
+  public synchronized void terminateTest(String testID) {
 
-    // TODO - should go into a stateless queue (so that we can recover)
-    Future<TerminationCriteriaResult> future = taskExecutorService.submit(task);
-
-    // replace with new task
-    testTasks.put(testID, future);
+    logger.info("terminateTest: " + testID);
 
     try {
 
-      TerminationCriteriaResult result = future.get();
-
-      switch (result) {
+      BenchFlowTestState testState = testModelDAO.getTestState(testID);
 
-        case GOAL_NOT_REACHABLE:
-          testModelDAO.setTestState(testID, TERMINATED);
-          testModelDAO.setTestTerminatedState(testID, COMPLETED_WITH_FAILURE);
+      switch (testState) {
+        case READY:
+          // remove from ready queue
+          readyQueue.remove(testID);
```

Comment:
I'm not sure I understand fully. It is only added in the `handleTestSate` method when the test is in state `READY`. This method is `synchronized` as well as the method here where it is removed.